reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+    }
+
+    if (this._ownerReadableByteStream === undefined) {
+      return;
+    }
+
+    if (this._readIntoRequests.length > 0) {
+      throw new TypeError('Tried to release a reader lock when that reader has pending read() calls un-settled');
+    }
+
+    if (this._ownerReadableByteStream._state === 'readable') {
+      CloseReadableByteStreamReader(this);
+    }
+  }
+}
+
+function CallPull(controller) {
+  const source = controller._underlyingByteSource;
+
+  const pullFunction = source['pull'];
+  if (pullFunction === undefined) {
+    return;
+  }
+
+  const stream = controller._controlledReadableByteStream;
+
+  if (typeof pullFunction !== 'function') {
+    DestroyReadableByteStreamController(controller);
+    ErrorReadableByteStream(stream, new TypeError('pull property of an underlying byte source must be a function'));
+    return;
+  }
+
+  controller._callPullOrPullIntoLaterIfNeeded = false;
+  controller._insideUnderlyingByteSource = true;
+
+  try {
+    pullFunction.apply(source);
+  } catch (e) {
+    DestoryReadableByteStreamController(controller);
+    if (stream._state === 'readable') {
+      ErrorReadableByteStream(stream, e);
+    }
+  }
+
+  controller._insideUnderlyingByteSource = false;
+}
+
+function CallPullInto(controller) {
+  const source = controller._underlyingByteSource;
+
+  const pullIntoFunction = source['pullInto'];
+  if (pullIntoFunction === undefined) {
+    return;
+  }
+
+  const stream = controller._controlledReadableByteStream;
+
+  if (typeof pullIntoFunction !== 'function') {
+    DestroyReadableByteStreamController(controller);
+    ErrorReadableByteStream(stream, new TypeError('pullInto property of an underlying byte source must be a function'));
+    return;
+  }
+
+  assert(controller._pendingPullIntos.length > 0);
+  const pullIntoDescriptor = controller._pendingPullIntos[0];
+
+  controller._callPullOrPullIntoLaterIfNeeded = false;
+  controller._insideUnderlyingByteSource = true;
+
+  try {
+    pullIntoFunction.apply(source, [pullIntoDescriptor.buffer,
+                                    pullIntoDescriptor.byteOffset + pullIntoDescriptor.bytesFilled,
+                                    pullIntoDescriptor.byteLength - pullIntoDescriptor.bytesFilled]);
+  } catch (e) {
+    DestroyReadableByteStreamController(controller);
+    const stream = controller._controlledReadableByteStream;
+    if (stream._state === 'readable') {
+      ErrorReadableByteStream(stream, e);
+    }
+  }
+
+  controller._insideUnderlyingByteSource = false;
+}
+
+function CallPullOrPullIntoLaterIfNeeded(controller) {
+  if (controller._insideUnderlyingByteSource) {
+    controller._callPullOrPullIntoLaterIfNeeded = true;
+    return;
+  }
+
+  Promise.resolve().then(CallPullOrPullIntoRepeatedlyIfNeeded.bind(undefined, controller));
+}
+
+function CallPullOrPullIntoRepeatedlyIfNeeded(controller) {
+  const stream = controller._controlledReadableByteStream;
+
+  while (true) {
```

Comment:
We can also choose to enqueue tasks to do that instead of while loop. But I just tried to repeat it synchronously for performance.